packages/block-library/src/code/utils.js

```diff
@@ -0,0 +1,117 @@
+ * @return {string} The given content with its ampersands converted into
+ *                  their HTML entity counterpart (i.e. & => &amp;)
+ */
+function escapeAmpersands( content ) {
+	return content.replace( /&/g, '&amp;' );
+}
+
+/**
+ * Returns the given content with all &amp; HTML entities converted into &.
+ *
+ * @param {string}  content The content of a code block.
+ * @return {string} The given content with all &amp; HTML entities
+ *                  converted into &.
+ */
+function unescapeAmpersands( content ) {
+	return content.replace( /&amp;/g, '&' );
+}
+
+/**
+ * Returns the given content with all opening shortcode characters converted
+ * into their HTML entity counterpart (i.e. [ => &#91;). For instance, a
+ * shortcode like [embed] becomes &#91;embed]
+ *
+ * This function replicates the escaping of HTML tags, where a tag like
+ * <strong> becomes &lt;strong>.
+ *
+ * @param {string}  content The content of a code block.
+ * @return {string} The given content with its opening shortcode characters
+ *                  converted into their HTML entity counterpart
+ *                  (i.e. [ => &#91;)
+ */
+function escapeOpeningSquareBrackets( content ) {
+	return content.replace( /\[/g, '&#91;' );
+}
+
+/**
+ * Returns the given content translating all &#91; into [.
+ *
+ * @param {string}  content The content of a code block.
+ * @return {string} The given content with all &#91; into [.
+ */
+function unescapeOpeningSquareBrackets( content ) {
+	return content.replace( /&#91;/g, '[' );
+}
+
+/**
+ * Converts the first two forward slashes of any isolated URL into their HTML
+ * counterparts (i.e. // => &#47;&#47;). For instance, https://youtube.com/watch?x
+ * becomes https:&#47;&#47;youtube.com/watch?x.
+ *
+ * An isolated URL is a URL that sits in its own line, surrounded only by spacing
+ * characters.
+ *
+ * See https://github.com/WordPress/wordpress-develop/blob/5.1.1/src/wp-includes/class-wp-embed.php#L403
+ *
+ * @param {string}  content The content of a code block.
+ * @return {string} The given content with its ampersands converted into
+ *                  their HTML entity counterpart (i.e. & => &amp;)
+ */
+function escapeProtocolInIsolatedUrls( content ) {
+	return content.replace( /^(\s*https?:)\/\/([^\s<>"]+\s*)$/m, '$1&#47;&#47;$2' );
+}
+
+/**
+ * Converts the first two forward slashes of any isolated URL from the HTML entity
+ * &#73; into /.
+ *
+ * An isolated URL is a URL that sits in its own line, surrounded only by spacing
+ * characters.
+ *
+ * See https://github.com/WordPress/wordpress-develop/blob/5.1.1/src/wp-includes/class-wp-embed.php#L403
+ *
+ * @param {string}  content The content of a code block.
+ * @return {string} The given content with the first two forward slashes of any
+ *                  isolated URL from the HTML entity &#73; into /.
+ */
+function unescapeProtocolInIsolatedUrls( content ) {
+	return content.replace( /^(\s*https?:)&#47;&#47;([^\s<>"]+\s*)$/m, '$//$2' );
```

Comment:
I don't think this works as expected. Namely, I'm guessing `$` should be `$1`.

```js
'https:&#47;&#47;www.youtube.com/watch?v=YuOBzWF0Aws'.replace( /^(\s*https?:)&#47;&#47;([^\s<>"]+\s*)$/m, '$//$2' );
// "$//www.youtube.com/watch?v=YuOBzWF0Aws"
```

```suggestion
	return content.replace( /^(\s*https?:)&#47;&#47;([^\s<>"]+\s*)$/m, '$1//$2' );
```

On that note, are you able to take a pass at including some unit tests for these functions?

Reply:
Thanks for catching that bug! I fixed it in d86123aea177254a8ee5996cbe834c7ace169efc.

> On that note, are you able to take a pass at including some unit tests for these functions?

Sure, I'll give it a shot and commit them later on today.

Reply:
> Sure, I'll give it a shot and commit them later on today.

No rush from my end, but just curious if you'd still planned to make these revisions?

Reply:
Oops, sorry! I tried to create the tests, but I don't know where they should be... I've seen some blocks have a `test` folder, but have no idea what's the correct approach. Any tips?

Reply:
> Oops, sorry! I tried to create the tests, but I don't know where they should be... I've seen some blocks have a `test` folder, but have no idea what's the correct approach. Any tips?

The general pattern is to create a file with the same name under a `test` directory relative to the file under test.

Example:

```
packages/block-library/src/code/utils.js
packages/block-library/src/code/test/utils.js
```

This could be better documented 😄

Reply:
Cool! I created a few unit tests in commit 1cb6ff810371f1f621b1cf19cb8a958d136da5cd. Oh, and I then renamed the test file from `index.js` to `utils.js` in 03e618dc6c547517a9b154313c8642199eac879a.

Is there anything left?

Reply:
Fixed in 1cb6ff810371f1f621b1cf19cb8a958d136da5cd and 03e618dc6c547517a9b154313c8642199eac879a.